typing/typeopt.ml

```diff
@@ -83,22 +83,23 @@ type classification =
   | Addr  (* anything except a float or a lazy *)
   | Any
 
-let classify env ty =
+let classify env ty : classification =
   let ty = scrape_ty env ty in
   if maybe_pointer_type env ty = Immediate then Int
   else match get_desc ty with
   | Tvar _ | Tunivar _ ->
       Any
   | Tconstr (p, _args, _abbrev) ->
-      if Path.same p Predef.path_float then Float
-      else if Path.same p Predef.path_lazy_t then Lazy
-      else if Path.same p Predef.path_string
-           || Path.same p Predef.path_bytes
-           || Path.same p Predef.path_array
-           || Path.same p Predef.path_nativeint
-           || Path.same p Predef.path_int32
-           || Path.same p Predef.path_int64 then Addr
-      else begin
+      begin match Predef.find_type_constr p with
+      | Some `Float -> Float
+      | Some `Lazy_t -> Lazy
+      | Some (`Int | `Char) -> Int
+      | Some (`String | `Bytes
+             | `Int32 | `Int64 | `Nativeint
+             | `Extension_constructor | `Continuation
+             | `Array | `Floatarray)
+        -> Addr
+      | Some #Predef.data_type_constr | None ->
```

Comment:
Since the data_type_constr are not abstracts by constructions, they can be moved to the branch above.

Reply:
You are correct, and I started to make the change that you propose, but it is tricky/fragile and I'm not sure anymore it's a good idea. Having the case `| Some #Predef.data_type_const -> Addr` relies on the following invariants for correctness:

1. I didn't mess up the split between abstract_type_constr and data_type_constr in Predef, the data_* ones really are not [Type_abstract]. 
2. It's not true that all datatypes are Addr (Bool and Unit are datatypes but Int), but we already checked against immediate types above so the classification of the *remaining* datatypes is Addr.
3. The reason why the remaining datatypes are Addr is that the code below says so, and this relies on the fact that *currently* a variant datatype cannot be a float or a lazy. This could become wrong if we allow unboxed constructors later on (for lazy; for float we would also need to disallow flat float arrays), and then the code below would change and we would need to think about this case again, but it would still be okay in practice because the predefined data types are not too weird (until we add unboxed options, that is), so they will remain Addr even if other datatypes become weirder.

Those are three not-so-simple assumptions that we would be rely on if we made the change you suggested, and which we are not relying on today. The change also slightly improves performance (it saves an environment lookup in these cases), but the current state is performance-preserving compared to what we had before (which also didn't special-case those types), so I would not consider it an important factor.

Reply:
For the record, the modified code looked like this:

```ocaml
      | Some #Predef.data_type_constr ->
          (* Predefined datatypes are not [Type_abstract],
             so (by the logic in the None case below) they
             are classified as [Addr]. *)
          Addr
      | None ->
```

Reply:
I can agree with the idea of not relying of an accidental alignment of two separated definitions.